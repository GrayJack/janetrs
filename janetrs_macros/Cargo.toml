[package]
name = "janetrs_macros"
version = "0.4.6"
authors = ["Eric Shimizu Karbstein <gr41.j4ck@gmail.com>"]
description = "Attribute macros for JanetRS"
repository = "https://github.com/GrayJack/janetrs"
homepage = "https://crates.io/crates/janetrs_macros"
documentation = "https://docs.rs/janetrs_macros"
license-file = "../LICENSE"
edition = "2021"
rust-version = "1.56.0"
keywords = ["janet", "embedding", "proc-macro"]
categories = ["development-tools::procedural-macro-helpers"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
name = "janetrs_macros"
path = "src/lib.rs"
proc-macro = true

[dependencies]
syn = { version = "1", features = ["extra-traits", "full"] }
quote = "1"
proc-macro2 = "1"
janetrs_version = "0.1"

[features]
default = []
system = ["janetrs_version/system"]

[[test]]
name = "tests"
path = "tests/progress.rs"

[dev-dependencies]
trybuild = { version = "1.0", features = ["diff"] }
janetrs = { path = "../../janetrs" }
