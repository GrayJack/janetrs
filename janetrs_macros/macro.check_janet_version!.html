<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.check_janet_version.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.check_janet_version.html">macro.check_janet_version.html</a>...</p>
    <script>location.replace("macro.check_janet_version.html" + location.search + location.hash);</script>
</body>
</html>