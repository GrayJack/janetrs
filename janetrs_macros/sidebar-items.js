window.SIDEBAR_ITEMS = {"attr":["cjvg","janet_fn","janet_version"],"macro":["check_janet_version","declare_janet_mod"]};