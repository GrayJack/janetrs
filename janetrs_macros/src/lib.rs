extern crate proc_macro;

use quote::{quote, quote_spanned};
use syn::{parse_macro_input, spanned::Spanned};

use janetrs_version::JanetVersion;

mod utils;
use utils::{janet_path_checker, Arg, Args, ArityArgs};

/// Macro that tranforms a high-level Janet function (`fn(&mut [Janet]) -> Janet`)
/// to the thing the Janet C API is expecting (`fn(i32, *mut janetrs::lowlevel::Janet) ->
/// janetrs::lowlevel::Janet`)
///
///
/// The optional argument `arity` adds a arity check for the function. It must receive the
/// kind of arity check. These kinds are `fix`, for fixed arity, and `range`, for ranged
/// or variadic arity. The `fix` kind receives a integer of the number of the parametes
/// the Janet function must have and the `range` kind can receive two arguments, the first
/// one if mandatory while the second one is optional, the first represents the minimal of
/// arguments the Janet function have to receive and the second represents the maximum of
/// arguments the Janet function can receive. If the maximum is not set for the range
/// arity, the maximum check is disabled, allowing variadic arguments.
///
///
/// The optional arg `check_mut_ref` adds a check to see if the function received more
/// than one reference to the same `*mut` pointer. This check is not the default because
/// Janet Types act like types with interior mutability and the check is expensive, but if
/// you want to make sure that your function never receives the same pointer more than
/// once you can use this.
///
///
/// **Usages**:
/// - `#[janet_fn]`
/// - `#[janet_fn(arity(fix(<N>)))]` where `N` is an integer literal
/// - `#[janet_fn(arity(range(<MIN> [, MAX])))]` where `MIN` and `MAX` are integer
///   literals
/// - `#[janet_fn(check_mut_ref)]`
/// - `#[janet_fn(arity(<...>), check_mut_ref)]` Combining both
#[proc_macro_attribute]
pub fn janet_fn(
    args: proc_macro::TokenStream, input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let func = parse_macro_input!(input as syn::Item);

    let args = parse_macro_input!(args as Args);

    let extra = args.0.iter().map(|arg| match arg {
        Arg::CheckMutRef => quote! {
            for j1 in &args[0..args.len() - 1] {
                for j2 in &args[1..] {
                    if matches!(
                        j1.kind(),
                        ::janetrs::JanetType::Array
                            | ::janetrs::JanetType::Buffer
                            | ::janetrs::JanetType::CFunction
                            | ::janetrs::JanetType::Fiber
                            | ::janetrs::JanetType::Function
                            | ::janetrs::JanetType::Keyword
                            | ::janetrs::JanetType::Pointer
                            | ::janetrs::JanetType::Symbol
                            | ::janetrs::JanetType::Table
                    ) && matches!(
                        j1.kind(),
                        ::janetrs::JanetType::Array
                            | ::janetrs::JanetType::Buffer
                            | ::janetrs::JanetType::CFunction
                            | ::janetrs::JanetType::Fiber
                            | ::janetrs::JanetType::Function
                            | ::janetrs::JanetType::Keyword
                            | ::janetrs::JanetType::Pointer
                            | ::janetrs::JanetType::Symbol
                            | ::janetrs::JanetType::Table
                    ) && j1 == j2
                    {
                        ::janetrs::jpanic!("Received two mutable references as arguments");
                    }
                }
            }
        },
        Arg::Arity(ArityArgs::Fix(n)) => quote! {
            ::janetrs::util::check_fix_arity(args, #n);
        },
        Arg::Arity(ArityArgs::Range(min, None)) => quote! {
            ::janetrs::util::check_range_arity(args, #min, None);
        },
        Arg::Arity(ArityArgs::Range(min, Some(max))) => quote! {
            ::janetrs::util::check_range_arity(args, #min, Some(#max));
        },
    });

    let ts = if let syn::Item::Fn(f) = func {
        let f_clone = f.clone();
        let attrs = f.attrs;
        let vis = f.vis;
        let name = f.sig.ident;
        // let syn::ReturnType::Type(_, type_path) = f.sig.output;

        // check for inputs
        if matches!(f.sig.inputs.len(), 1) {
            if let Some(syn::FnArg::Typed(syn::PatType { ty, .. })) = f.sig.inputs.first() {
                if let syn::Type::Reference(syn::TypeReference {
                    mutability,
                    ref elem,
                    ..
                }) = **ty
                {
                    if mutability.is_none() {
                        return quote_spanned! {ty.span() => compile_error!("expected argument to be a mutable reference and found a immutable reference");}.into();
                    } else if let syn::Type::Slice(syn::TypeSlice {
                        elem: ref slice, ..
                    }) = **elem
                    {
                        if let syn::Type::Path(syn::TypePath { ref path, .. }) = **slice {
                            if !janet_path_checker(path) {
                                return quote_spanned! {path.span() => compile_error!("expected to be a `janetrs::Janet`");}.into();
                            }
                        } else {
                            return quote_spanned! {slice.span() => compile_error!("expected to be a `janetrs::Janet`");}.into();
                        }
                    } else {
                        return quote_spanned! {elem.span() => compile_error!("expected to be a slice of `janetrs::Janet`");}.into();
                    }
                } else {
                    return quote_spanned! {ty.span() => compile_error!("expected argument to be a mutable reference and found something that is not a reference at all");}.into();
                }
            }

            // check output type
            let output_span = match f.sig.output.clone() {
                syn::ReturnType::Default => f_clone.sig.span(),
                syn::ReturnType::Type(_, ty) => ty.span(),
            };
            if let syn::ReturnType::Type(_, ty) = f.sig.output {
                if let syn::Type::Path(syn::TypePath { ref path, .. }) = *ty {
                    if !janet_path_checker(path) {
                        return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::Janet`");}.into();
                    }
                } else {
                    return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::Janet`");}.into();
                }
            } else {
                return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::Janet`");}.into();
            }

            quote! {
                #(#attrs)*
                #[no_mangle]
                #vis unsafe extern "C" fn #name(argc: i32, argv: *mut ::janetrs::lowlevel::Janet) -> ::janetrs::lowlevel::Janet {
                    #[inline]
                    #f_clone

                    let args = unsafe { core::slice::from_raw_parts_mut(argv, argc as usize) };
                    let mut args = unsafe { &mut *(args as *mut [::janetrs::lowlevel::Janet] as *mut [::janetrs::Janet])};

                    #(#extra)*

                    #name(args).into()
                }
            }
        } else {
            quote_spanned! {f_clone.sig.inputs.span() => compile_error!("expected exactly one argument of type `&mut [janetrs::Janet]`");}
        }
    } else {
        quote_spanned! {func.span() => compile_error!("expected fn item");}
    };

    ts.into()
}


const CURRENT_JANET: JanetVersion = JanetVersion::current();

/// Conditional Janet Version Gate
///
/// **Usage:** `#[janet_version(<MIN_VERSION>, [MAX_VERSION])]` where `MIN_VERSION` and
/// `MAX_VERSION` are string literals.
///
/// A macro da conditionally includes the `input` if the version of Janet is bigger or
/// equal to the passed minimal version and smaller than the passed maximun version.
///
/// That means that the range is open in the maximum version: [MIN_VERSION, MAX_VERSION).
#[proc_macro_attribute]
pub fn janet_version(
    args: proc_macro::TokenStream, input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let args = parse_macro_input!(args as syn::AttributeArgs);
    // let input = parse_macro_input!(input as syn::Item);

    if args.len() > 2 {
        return quote!{compile_error!("expected at max two argument to the janet_version proc-macro");}
            .into();
    }

    if args.is_empty() {
        return quote!{compile_error!("expected at least one argument to the janet_version proc-macro");}
            .into();
    }

    let min_lit = if let syn::NestedMeta::Lit(syn::Lit::Str(ref lit)) = args[0] {
        lit
    } else {
        return quote_spanned! {args[0].span() => compile_error!("the argument must be a string literal");}.into();
    };

    let max_lit = match args.get(1) {
        Some(syn::NestedMeta::Lit(syn::Lit::Str(ref lit))) => Some(lit),
        None => None,
        _ => return quote_spanned! {args[1].span() => compile_error!("the argument must be a string literal");}.into()
    };

    match parse_args(&min_lit.value()) {
        Ok(req_min_ver) => {
            if let Some(max_lit) = max_lit {
                match parse_args(&max_lit.value()) {
                    Ok(req_max_ver) => {
                        if req_min_ver <= CURRENT_JANET && req_max_ver > CURRENT_JANET {
                            input
                        } else {
                            proc_macro::TokenStream::new()
                        }
                    },
                    Err(err) => {
                        let err = format!("invalid string literal: {}", err);
                        (quote_spanned! {max_lit.span() => compile_error!(#err);}).into()
                    },
                }
            } else if req_min_ver <= CURRENT_JANET {
                input
            } else {
                proc_macro::TokenStream::new()
            }
        },
        Err(err) => {
            let err = format!("invalid string literal: {}", err);
            (quote_spanned! {min_lit.span() => compile_error!(#err);}).into()
        },
    }
}

fn parse_args(arg: &str) -> Result<JanetVersion, String> {
    let vec_values = arg
        .split('.')
        .map(|e| e.parse::<u32>())
        .collect::<Result<Vec<_>, _>>()
        .map_err(|err| format!("{}", err))?;

    Ok(match &vec_values[..] {
        [major] => JanetVersion::custom(*major, 0, 0),
        [major, minor] => JanetVersion::custom(*major, *minor, 0),
        [major, minor, patch] => JanetVersion::custom(*major, *minor, *patch),
        _ => return Err("Invalid version string".to_string()),
    })
}

/// Conditional Janet Version Gate
///
/// **Usage:** `#[cjvg(<MIN_VERSION>, [MAX_VERSION])]` where `MIN_VERSION` and
/// `MAX_VERSION` are string literals.
///
/// A macro da conditionally includes the `input` if the version of Janet is bigger or
/// equal to the passed minimal version and smaller than the passed maximun version.
///
/// That means that the range is open in the maximum version: [MIN_VERSION, MAX_VERSION).
#[proc_macro_attribute]
pub fn cjvg(
    args: proc_macro::TokenStream, input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    janet_version(args, input)
}
