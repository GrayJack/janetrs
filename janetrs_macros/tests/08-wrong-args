
use janetrs_macros::janet_fn;

#[janet_fn(a)]
pub fn rust_hello(_args: &mut [Janet]) -> Janet {
    println!("Hello from Rust!");
    Janet::nil()
}

#[janet_fn(arity(fix("")))]
pub fn test1(_args: &mut [Janet]) -> Janet {
    Janet::nil()
}

#[janet_fn(arity(fix))]
pub fn test2(_args: &mut [Janet]) -> Janet {
    Janet::nil()
}

#[janet_fn(arity(fix()))]
pub fn test3(_args: &mut [Janet]) -> Janet {
    Janet::nil()
}

#[janet_fn(arity(range()))]
pub fn test4(_args: &mut [Janet]) -> Janet {
    Janet::nil()
}

#[janet_fn(check_mut_ref, check_mut_ref)]
pub fn test5(_args: &mut [Janet]) -> Janet {
    Janet::nil()
}

#[janet_fn(arity(fix(1)), check_mut_ref, arity(fix(2)))]
pub fn test6(_args: &mut [Janet]) -> Janet {
    Janet::nil()
}

fn main() {}