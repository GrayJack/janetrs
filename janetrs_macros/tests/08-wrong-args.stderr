error: Invalid argument for the macro. Expected `arity` or `check_mut_ref`
 --> $DIR/08-wrong-args:4:12
  |
4 | #[janet_fn(a)]
  |            ^

error: expected integer literal
  --> $DIR/08-wrong-args:10:22
   |
10 | #[janet_fn(arity(fix("")))]
   |                      ^^

error: unexpected end of input, expected parentheses
  --> $DIR/08-wrong-args:15:21
   |
15 | #[janet_fn(arity(fix))]
   |                     ^

error: unexpected end of input, expected integer literal
  --> $DIR/08-wrong-args:20:22
   |
20 | #[janet_fn(arity(fix()))]
   |                      ^

error: Invalid number of arguments for `range`: Expected at least 1, with max of 2 arguments, got 0
  --> $DIR/08-wrong-args:25:24
   |
25 | #[janet_fn(arity(range()))]
   |                        ^

error: Repeated argument kind: There must be only one argument of each kind, that is, only one of `arity` or `check_mut_ref`
  --> $DIR/08-wrong-args:30:1
   |
30 |   #[janet_fn(check_mut_ref, check_mut_ref)]
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ in this macro invocation
   |
  ::: $WORKSPACE/janetrs_macros/src/lib.rs
   |
   | / pub fn janet_fn(
   | |     args: proc_macro::TokenStream, input: proc_macro::TokenStream,
   | | ) -> proc_macro::TokenStream {
   | |____________________________- in this expansion of procedural macro `#[janet_fn]`

error: expected a maximum of two arguments to the janet_fn proc-macro
  --> $DIR/08-wrong-args:35:1
   |
35 |   #[janet_fn(arity(fix(1)), check_mut_ref, arity(fix(2)))]
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ in this macro invocation
   |
  ::: $WORKSPACE/janetrs_macros/src/lib.rs
   |
   | / pub fn janet_fn(
   | |     args: proc_macro::TokenStream, input: proc_macro::TokenStream,
   | | ) -> proc_macro::TokenStream {
   | |____________________________- in this expansion of procedural macro `#[janet_fn]`
